### PR TITLE
init bridge monitoring tests

### PR DESCRIPTION
**Pull request by Omarabdul3ziz**
### Related Issues
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/383

image

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstring

### CODE REVIEW

File: monitoring-bot/internal/constants.go [hunk 1]
```diff
@@ -6,10 +6,21 @@ const (
 	testNetwork network = "testnet"
 	devNetwork  network = "devnet"
 	qaNetwork   network = "qanet"
+
+	tftIssuerAddress    = "GA47YZA3PKFUZMPLQ3B5F2E3CJIB57TGGU7SPCQT2WAEYKN766PWIMB3"
```
Comment (xmonader) [inline monitoring-bot/internal/constants.go]: the address is different on networks

File: monitoring-bot/README.md [hunk 2]
```diff
@@ -3,93 +3,109 @@
 <a href='https://github.com/jpoles1/gopherbadger' target='_blank'>![gopherbadger-tag-do-not-edit](https://img.shields.io/badge/Go%20Coverage-48%25-brightgreen.svg?longCache=true&style=flat)</a>
 [![Testing](https://github.com/threefoldtech/tfgrid-sdk-go/actions/workflows/monbot-test.yml/badge.svg?branch=development_mono)](https://github.com/threefoldtech/tfgrid-sdk-go/actions/workflows/monbot-test.yml) [![Testing](https://github.com/threefoldtech/tfgrid-sdk-go/actions/workflows/monbot-lint.yml/badge.svg?branch=development_mono)](https://github.com/threefoldtech/tfgrid-sdk-go/actions/workflows/monbot-lint.yml)
 
-This is a bot to monitor the balance in accounts and send warnings if it is under some threshold.
-It also monitors the behavior of the rmb proxy
+This is a bot to monitor some TFGrid functionalities here is a list:
+
+- liveness of gridproxy on different networks and rmb call behavior to some farmer nodes.
+- the balance in accounts and send warnings if it is under some threshold.
```
Comment (rawdaGastan) [inline monitoring-bot/README.md:9]: add bridge report

File: monitoring-bot/README.md [hunk 3]
```diff
@@ -3,93 +3,109 @@
...
+    DEVNET_MNEMONIC=<your devnet mnemonic>
+    QANET_MNEMONIC=<your qanet mnemonic>
+    DEV_FARM_NAME=Freefarm
+    QA_FARM_NAME=Freefarm
+    MAIN_FARM_NAME=Freefarm
+    TEST_FARM_NAME=FreeFarm
+    BOT_TOKEN=<bot token. you got it after creating the bot>
+    CHAT_ID=<your personal chat ID, where bot will send you >
+    MINS=<number of minutes between each message>
+    BRIDGE_MON_INTERVAL_MIN=<number of minutes between each bridge test report>
```
Comment (rawdaGastan) [inline monitoring-bot/README.md]: I think we can use mins?

File: monitoring-bot/internal/bridge.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+)
+
+func (m *Monitor) monitorBridges() error {
+	message := "TfChain Bridges to Stellar Monitor Results:\n"
+	for _, net := range networks {
+
+		res, err := m.monitorBridge(net)
+		if err != nil {
+			message += fmt.Sprintf("\n - %s\n%s\n\n", res, err.Error())
+			fmt.Println(err.Error())
```
Comment (rawdaGastan) [inline monitoring-bot/internal/bridge.go]: can we use zerolog instead?

File: monitoring-bot/internal/bridge.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	if err := m.sendBotMessage(message); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *Monitor) monitorBridge(net network) (string, error) {
```
Comment (rawdaGastan) [inline monitoring-bot/internal/bridge.go]: I think we should return the error and check which msg to report in `monitorBridges`

File: monitoring-bot/internal/bridge.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+	err = m.sendToStellar(net)
+	if err != nil {
+		return failureMessage, err
+	}
+
+	return fmt.Sprintf("Bridge for %v is working ✅", net), nil
+}
+
+func (m *Monitor) sendToTfChain(tfChain network) error {
```
Comment (rawdaGastan) [inline monitoring-bot/internal/bridge.go]: `tfChain` should be net?

File: monitoring-bot/internal/bridge.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+	defer conn.Close()
+
+	identity, err := client.NewIdentityFromSr25519Phrase(m.mnemonics[tfChain])
+	if err != nil {
+		return fmt.Errorf("failed to get identity from mnemonic: %w", err)
+	}
+
+	twinID, err := conn.GetTwinByPubKey(identity.PublicKey())
+	if err != nil {
+		return fmt.Errorf("failed to get twinId: %w", err)
```
Comment (rawdaGastan) [inline monitoring-bot/internal/bridge.go]: `twinId` should be `twin id` (because it's an uppercase char)

File: monitoring-bot/internal/bridge.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+	_, err = horizonclient.DefaultTestNetClient.SubmitTransaction(tx)
+	if err != nil {
+		return fmt.Errorf("failed to submit the txn: %w", err)
+	}
+
+	return nil
+}
+
+func (m *Monitor) sendToStellar(tfChain network) error {
```
Comment (rawdaGastan) [inline monitoring-bot/internal/bridge.go]: tfChain ->  net?

File: monitoring-bot/internal/bridge.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+	strSecret := m.env.testStellarSecret
+	stellarTFTIssuerAddress := tftIssuerStellarTest
+	if tfChain == mainNetwork || tfChain == testNetwork {
+		strSecret = m.env.publicStellarSecret
+		stellarTFTIssuerAddress = tftIssuerStellarPublic
+	}
+
+	tftTrustLine := txnbuild.CreditAsset{Code: "TFT", Issuer: stellarTFTIssuerAddress}
+	strClient := horizonclient.DefaultTestNetClient
```
Comment (rawdaGastan) [inline monitoring-bot/internal/bridge.go:121]: is it for test and public?

File: monitoring-bot/internal/monitor.go [hunk 10]
```diff
@@ -213,25 +228,27 @@ func (m *Monitor) sendMessage(manager client.Manager, wallet wallet) error {
...
 		"application/json",
 		bytes.NewBuffer(body),
 	)
-	if err != nil {
-		return err
-	}
-	if response.StatusCode >= 400 {
-		return errors.New("request send message failed")
+
+	if err != nil || res.StatusCode >= 400 {
```
Comment (rawdaGastan) [inline monitoring-bot/internal/monitor.go]: there is no error for the second case

**Review by rawdaGastan — CHANGES_REQUESTED** [review]

File: monitoring-bot/internal/monitor.go [hunk 11]
```diff
@@ -194,44 +198,60 @@ func (m *Monitor) Start() {
...
-	response, err := http.Post(
+
+	res, err := http.Post(
 		url,
 		"application/json",
 		bytes.NewBuffer(body),
 	)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to send http post request: %w", err)
```
Comment (rawdaGastan) [inline monitoring-bot/internal/monitor.go]: can we make error msgs -> `failed to send telegram message request` 
- for res.StatusCode >= 400 also

File: monitoring-bot/internal/monitor.go [hunk 12]
```diff
@@ -194,44 +198,60 @@ func (m *Monitor) Start() {
...
 
 // getTelegramUrl returns the telegram bot api url
 func (m *Monitor) getTelegramURL() string {
-	return fmt.Sprintf("https://api.telegram.org/bot%s", m.env.botToken)
+	return fmt.Sprintf("%s%s", telegramBotURL, m.env.botToken)
 }
 
 // sendMessage sends a message with the balance to a telegram bot
 // if it is less than the tft threshold
 func (m *Monitor) sendMessage(manager client.Manager, wallet wallet) error {
```
Comment (rawdaGastan) [inline monitoring-bot/internal/monitor.go]: can we rename func 
for example `balanceAlert`

File: monitoring-bot/internal/bridge.go [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+	identity, err := client.NewIdentityFromSr25519Phrase(m.mnemonics[net])
+	if err != nil {
+		return balanceReport{}, err
+	}
+
+	// get current balance
+	originalBalance, err := m.getBalance(conn, address(identity.Address()))
+	if err != nil {
+		return balanceReport{}, fmt.Errorf("failed to get balance for account ")
```
Comment (rawdaGastan) [inline monitoring-bot/internal/bridge.go]: you didn't wrap the error

File: monitoring-bot/internal/bridge.go [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+	return func(conn *client.Substrate, identity client.Identity, net network) (float64, error) {
+		if err := tx(conn, identity, net); err != nil {
+			return 0, err
+		}
+
+		<-time.After(balanceWaitIntervalSeconds * time.Second)
+
+		balanceAfterChain, err := m.getBalance(conn, address(identity.Address()))
+		if err != nil {
+			return 0, fmt.Errorf("failed to get balance for account ")
```
Comment (rawdaGastan) [inline monitoring-bot/internal/bridge.go]: you didn't wrap the error

**Review by rawdaGastan — CHANGES_REQUESTED** [review]

**Comment by rawdaGastan** [comment]
can you please update the report msg in the description?

**Review by rawdaGastan — APPROVED** [review]

**Pull request merged**